_SPIKE_.md

```diff
@@ -0,0 +1,163 @@
+   Extension/ExtensionPoint pattern and delegates the remaining work to the
+   plugin.
+
+3. An official model-api-builder plugin for building CRUD REST APIs using
+   `DefaultCrudRepository` implementation. The plugin is implemented inside the
+   recently introduced package `@loopback/rest-crud`.
+
+Under the hood, the CRUD REST API builder performs two steps:
+
+1. It defines a model-specific repository class, e.g. `ProductRepository` and
+   registers it with the Application context for Dependency Injection. In the
+   future, we can make this step optional and allow app developers to supply
+   their own repository class (e.g. as created by `lb4 repository`) 💪
+
+   The repository class is decorated to receive the dataSource instance via DI.
+
+2. Then a model-specific controller class is defined, e.g. `ProductController`
+   and decorated to receive the repository instance via DI. The controller is
+   registered with the Application context to get it included in the public API.
+
+I feel it's important to use model-named classes (e.g. `ProductController`)
+instead of reusing the same controller/repository class name for all models.
+Model-named classes make error troubleshooting and debugging easier, because
+stack traces include model name in function names. (Compare
+`ProductController.replaceById` with `CrudRestControllerImpl.replaceById` -
+which one is more useful?)
+
+In my proposal, model-config files are in JSON format to make programmatic edits
+easier. This has a downside in TypeScript projects - these config files must
+live outside `src` because TypeScript does not copy arbitrary JSON files.
+
+### Extensibility & customization options
+
+The proposed design enables the following opportunities to extend and customize
+the default behavior of API endpoints:
+
+- App developers can create & bind a custom repository class, this allows them
+  e.g. to implement functionality similar to LB3 Operation Hooks.
+
+- App developers can implement their own api-builder plugins, replacing the
+  repository & controller builders provided by LB4 by their own logic.
+
+- Model configuration schema is extensible, individual plugins can define
+  additional model-config options to further tweak the behavior of API
+  endpoints.
+
+## How to review the spike
+
+I have updated `examples/todo` application to leverage this new functionality,
+you can look around the updated code to see what the user experience will look
+like. Take a look at how acceptance level tests are accessing the repository
+created by `rest-crud` plugin.
+
+The booter is implemented in
+[`packages/booter-rest/src/rest.booter.ts`](https://github.com/strongloop/loopback-next/blob/spike/crud-rest-booter/packages/booter-rest/src/rest.booter.ts).
+Initially, I created a new package for this booter, but on the second thought, I
+think this booter is generic enough to be included in `@loopback/boot` under a
+different name `ModelApiBooter`.
```

Comment:
> Initially, I created a new package for this booter, but on the second thought, I
think this booter is generic enough to be included in `@loopback/boot` under a
different name `ModelApiBooter`.

+1 Or merge it with the `@loopback/model-api-builder` and export everything as a component.

Reply:
I don't want to merge it with `@loopback/model-api-builder` because I don't want `@loopback/model-api-builder` to be coupled with `@loopback/boot`.

`@loopback/model-api-builder` is going to be used a dependency of all builder implementations, I want to keep it as lightweight as possible.

Reply:
> because I don't want @loopback/model-api-builder to be coupled with @loopback/boot.

I see fair enough 👍